src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,325 @@
+        return QVariant();
+
+    const int row = index.internalId();
+    Q_ASSERT(isValidRow(row));
+    const TagModelItem &item = m_tagItems[row];
+
+    switch (role) {
+    case Qt::DecorationRole:
+        return GuiIconProvider::instance()->getIcon("inode-directory");
+    case Qt::DisplayRole:
+        return QString(QLatin1String("%1 (%2)"))
+               .arg(tagDisplayName(item.tag())).arg(item.torrentsCount());
+    case Qt::UserRole:
+        return item.torrentsCount();
+    default:
+        return QVariant();
+    }
+}
+
+Qt::ItemFlags TagFilterModel::flags(const QModelIndex &index) const
+{
+    if (!index.isValid())
+        return 0;
+    return Qt::ItemIsEnabled | Qt::ItemIsSelectable;
+}
+
+QVariant TagFilterModel::headerData(int section, Qt::Orientation orientation, int role) const
+{
+    if ((orientation == Qt::Horizontal) && (role == Qt::DisplayRole))
+        if (section == 0)
+            return tr("Tags");
+    return QVariant();
+}
+
+QModelIndex TagFilterModel::index(int row, int, const QModelIndex &) const
+{
+    if (!isValidRow(row))
+        return QModelIndex();
+    return createIndex(row, 0, row);
+}
+
+int TagFilterModel::rowCount(const QModelIndex &parent) const
+{
+    if (!parent.isValid())
+        return m_tagItems.count();
+    return 0;
+}
+
+bool TagFilterModel::isValidRow(int row) const
+{
+    return (row >= 0) && (row < m_tagItems.size());
+}
+
+QModelIndex TagFilterModel::index(const QString &tag) const
+{
+    const int row = findRow(tag);
+    if (!isValidRow(row))
+        return QModelIndex();
+    return index(row, 0, QModelIndex());
+}
+
+QString TagFilterModel::tag(const QModelIndex &index) const
+{
+    if (!index.isValid())
+        return QString();
+    const int row = index.internalId();
+    Q_ASSERT(isValidRow(row));
+    return m_tagItems[row].tag();
+}
+
+void TagFilterModel::tagAdded(const QString &tag)
+{
+    const int row = m_tagItems.count();
+    beginInsertRows(QModelIndex(), row, row);
+    addToModel(tag, 0);
+    endInsertRows();
+}
+
+void TagFilterModel::tagRemoved(const QString &tag)
+{
+    QModelIndex i = index(tag);
+    beginRemoveRows(i.parent(), i.row(), i.row());
+    removeFromModel(i.row());
+    endRemoveRows();
+}
+
+void TagFilterModel::torrentTagAdded(BitTorrent::TorrentHandle *const torrent, const QString &tag)
+{
+    if (torrent->tags().count() == 1)
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    const int row = findRow(tag);
+    Q_ASSERT(isValidRow(row));
+    TagModelItem &item = m_tagItems[row];
+
+    item.increaseTorrentsCount();
+    const QModelIndex i = index(row, 0, QModelIndex());
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentTagRemoved(BitTorrent::TorrentHandle* const torrent, const QString &tag)
+{
+    Q_ASSERT(torrent->tags().count() >= 0);
+    if (torrent->tags().count() == 0)
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
```

Comment:
Methods aren't const, since they return non-const member pointers, but otherwise done (here and elsewhere).